File: Sources/ATTNEventTracker.m
//
//  ATTNEventTracker.m
//  attentive-ios-sdk
//
//  Created by Wyatt Davis on 12/7/22.
//

#import "ATTNEventTracker.h"
#import "attentive_ios_sdk_framework/attentive_ios_sdk_framework-Swift.h"
#import "ATTNAPI.h"

static ATTNEventTracker* __sharedInstance = nil;

@implementation ATTNEventTracker {
  ObjcATTNSDK* _sdk;
}

+ (void)setupWithSdk:(ObjcATTNSDK*)sdk {
  static dispatch_once_t ensureOnlyOnceToken;
  dispatch_once(&ensureOnlyOnceToken, ^{
    __sharedInstance = [[self alloc] initWithSdk:sdk];
  });
}

- (id)initWithSdk:(ObjcATTNSDK*)sdk {
  if (self = [super init]) {
    _sdk = sdk;
  }

  return self;
}

- (void)recordEvent:(id<ATTNEvent>)event {
  // TODO: Would be good to clone the UserIdentity so any changes to UserIdentity from another thread don't interfere with the API code
  [[_sdk getApi] sendEvent:event userIdentity:[_sdk getUserIdentity]];
}

+ (instancetype)sharedInstance {
  NSAssert(__sharedInstance != nil, @"ATTNEventTracker must be setup before being used");
  return __sharedInstance;
}

@end

File: Sources/Public/ATTNEventTracker.swift
//
//  ATTNEventTracker.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-30.
//

import Foundation

@objc(ATTNEventTracker)
public final class ATTNEventTracker: NSObject {
  private static var _sharedInstance: ATTNEventTracker?
  private let sdk: ATTNSDK

  @objc(initWithSdk:)
  public init(sdk: ATTNSDK) {
    self.sdk = sdk
  }

  @objc(setupWithSdk:)
  public static func setup(with sdk: ATTNSDK) {
    _sharedInstance = ATTNEventTracker(sdk: sdk)
    Loggers.event.debug("ATTNEventTracker was initialized with SDK")
  }

  @available(swift, deprecated: 0.6, message: "Please use record(event: ATTNEvent) instead.")
  @objc(recordEvent:)
  public func record(_ event: ATTNEvent) {
    sdk.send(event: event)
  }

  public func record(event: ATTNEvent) {
    sdk.send(event: event)
  }

  @objc
  public static func sharedInstance() -> ATTNEventTracker? {
    assert(_sharedInstance != nil, "ATTNEventTracker must be setup before being used")
    return _sharedInstance
  }
}

// MARK: Internal Helpers
extension ATTNEventTracker {
  static func destroy() {
    _sharedInstance = nil
  }

  func getSdk() -> ATTNSDK {
    sdk
  }
}
